File: 1301-1400/1372-Longest ZigZag Path in a Binary Tree/1372-Longest ZigZag Path in a Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def longestZigZag(self, root: TreeNode) -> int:
        maxLen = 0
        def dfs(root):
            if root is None:
                return [-1, -1]
            left = dfs(root.left)
            right = dfs(root.right)
            nonlocal maxLen
            maxLen = max(maxLen, 1 + max(left[1], right[0]))
            return [1 + left[1], 1 + right[0]]
        dfs(root)
        return maxLen

File: 1301-1400/1372-Longest ZigZag Path in a Binary Tree/1372-Longest ZigZag Path in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int longestZigZag(TreeNode* root) {
        int maxLen = 0, leftLen = -1, rightLen = -1;
        dfs(root, leftLen, rightLen, maxLen);
        return maxLen;
    }
    
private:
    void dfs(TreeNode* root, int& leftLen, int& rightLen, int& maxLen) {
        if (root == nullptr) {
            return;
        }
        int leftLen1 = -1, rightLen1 = -1, leftLen2 = -1, rightLen2 = -1;
        dfs(root->left, leftLen1, rightLen1, maxLen);
        dfs(root->right, leftLen2, rightLen2, maxLen);
        maxLen = max(maxLen, 1 + max(rightLen1, leftLen2));
        leftLen = 1 + rightLen1;
        rightLen = 1 + leftLen2;
    }
};
